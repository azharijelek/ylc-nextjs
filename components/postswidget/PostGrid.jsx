import Link from 'next/link'
import PropTypes from 'prop-types'

/**
 * PostGrid
 * @param {*} props id, title, thumbnail, permalink
 */

export default function PostGrid(props) {
  let thumbnail = `${props.thumbnail}&h=140`
  let catLink = props.categories.slug
  let catName = props.categories.name
  return (
    <>
      <article className={'post-grid post-' + props.id}>
        <div className="post-grid--thumb">
          <Link href={catLink}>
            <a className="cat-badge">
              <span dangerouslySetInnerHTML={{ __html: catName }} />
            </a>
          </Link>
          <Link href={props.permalink}>
            <a>
              <img src={thumbnail} alt={props.title} width="60" height="60" loading="lazy" />
            </a>
          </Link>
        </div>
        <div className="postDetail">
          <h4>
            <Link href={props.permalink}>
              <a>{props.title}</a>
            </Link>
          </h4>
          <p>{props.blurb}</p>
        </div>
      </article>

      <style jsx>{`
        .post-grid {
          .post-grid--thumb {
            width: 100%;
            height: 140px;
            overflow: hidden;
            margin-bottom: 10px;
            overflow: hidden;
            background: #e0e0e0;
            border-radius: 8px;
            line-height: 0;
            position: relative;

            .cat-badge {
              position: absolute;
              top: 0;
              left: 0;
              height: 20px;
              line-height: 20px;
              background: #4f4f4f;
              color: #fff;
              display: inline-block;
              padding: 0 5px;
              font-size: 12px;
            }

            img {
              object-fit: cover !important;
              width: 100% !important;
              height: 100% !important;
            }
          }
          h4 {
            font-size: 18px;
            margin-top: 10px;
            line-height: 24px;
            margin-bottom: 10px;
          }
          p {
            margin-top: 0;
            font-size: 14px;
            line-height: 22px;
          }
        }
      `}</style>
    </>
  )
}

PostGrid.propTypes = {
  thumbnail: PropTypes.string,
  categories: PropTypes.object,
  permalink: PropTypes.string,
  title: PropTypes.string,
  id: PropTypes.number,
  blurb: PropTypes.string
}
