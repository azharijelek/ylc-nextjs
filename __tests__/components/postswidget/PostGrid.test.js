import React from 'react'
import { shallow, render } from 'enzyme'
import PostGrid from '@/components/postswidget/PostGrid'

describe('<PostGrid/>', () => {
  let wrapper, post, PostGridComp

  beforeEach(() => {
    post = {
      blurb: 'How to tell a good password manager from a bad one.',
      categories: { slug: '/technology/safety-online/secure-access/', name: 'Secure Access' },
      date: 'September 2, 2020',
      featured_img:
        'https://cdn.statically.io/img/yourlifechoices.s3-ap-southeast-2.amazonaws.com/old/assets/autouploads/f4aa5f507148.jpg?f=webp',
      id: 533593,
      permalink:
        'http://localhost:4200/technology/safety-online/secure-access/password-managers-explained/',
      slug: '/technology/safety-online/secure-access/password-managers-explained/',
      title: 'Password managers explained: do you need one? Are they safe?'
    }

    PostGridComp = (
      <PostGrid
        id={post.id}
        title={post.title}
        thumbnail={post.featured_img}
        permalink={post.permalink}
        categories={post.categories}
        blurb={post.blurb}
        key={'post-' + post.id}
      />
    )

    wrapper = shallow(PostGridComp)
  })

  it('should renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have article with class: .post-grid post-${post_id}', () => {
    expect(wrapper.find('article')).toHaveLength(1)
    expect(wrapper.find('.post-' + post.id)).toHaveLength(1)
  })

  it('should have post-grid--thumb', () => {
    expect(wrapper.find('.post-grid--thumb')).toHaveLength(1)
    expect(wrapper.find('.post-grid--thumb').find('img')).toHaveLength(1)
  })

  it('should have .cat-badge', () => {
    expect(wrapper.find('.cat-badge')).toHaveLength(1)
  })

  it('should have render category name correctly', () => {
    const tree = render(PostGridComp)
    const categorybadge = tree.find('.cat-badge')
    expect(categorybadge.find('span').text()).toBe(post.categories.name)
  })

  it('should have .postDetail class', () => {
    expect(wrapper.find('.postDetail')).toHaveLength(1)
  })

  it('should have .postDetail class', () => {
    expect(wrapper.find('.postDetail')).toHaveLength(1)
  })

  it('.postDetail should have <h4> and <a>', () => {
    expect(wrapper.find('h4')).toHaveLength(1)
    expect(wrapper.find('h4 a')).toHaveLength(1)
  })

  it('.postDetail anchor should has <Link> component', () => {
    expect(wrapper.find('h4 Link')).toHaveLength(1)
  })

  it('<Link> href should contain correct permalink', () => {
    expect(wrapper.find('h4 Link').prop('href')).toBe(post.permalink)
  })

  it('.postDetail anchor should render correct post name', () => {
    expect(wrapper.find('h4 a').text()).toBe(post.title)
  })

  it('.postDetail should have p tag and render correct blurb text', () => {
    expect(wrapper.find('.postDetail').find('p')).toHaveLength(1)
    expect(wrapper.find('.postDetail').find('p').text()).toBe(post.blurb)
  })
})
