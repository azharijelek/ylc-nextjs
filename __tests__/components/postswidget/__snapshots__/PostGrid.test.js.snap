// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostGrid/> should renders correctly 1`] = `ShallowWrapper {}`;
